Use pandoc instead of make4ht

gnuhead-chieb opened the pull request:
Please don't merge PR for now.This PR is in progress now.
I have to fix more issues and create translation process

gnuhead-chieb commented:
Milestones

- [X] Custom color name does not work -Fixed
- [X] Icon doesn't show correctly -Fixed
- [X] Alertbox doesn't show correctly -Fixed
- [X]Label hyperref broken in pages/app-details-page.tex and pages/settings-page.tex -Fixed
- [ ] Listing doesn't working in appendices/app_ops.tex(C.3.3=7.3.3)
- [ ] Add link on section title
- [ ] Hide level5 section number
- [ ] Add doctool.sh commanline controller
- [ ] Add string key for making translatable with xliff ( %%!! string_key<< STRING %%!!>>)
- [ ] Extract string and update xliff translation file
- [ ] Restore TEX from xliff

Alexisblues reviewed (approved):
```

```

gnuhead-chieb commented:
Does anyone have an idea to merge XML translation(content) file to Tex?
Source Tex Format
```
\(section|subsection|subsubsection){SECTION TITLE} %%##title_key==title>>
%%!!string_key<<
CONTENTCONTENTCONTENT
CONTENTCONTENTCONTENT
%%!!>>
```

XML Format
```
<?xml version="1.0" encoding="utf-8"?>
<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
<string name=title_key==title">SECTION TITLE</string>
...
<string name="faq_aot-usbdebugging">CONTENTCONTENTCONTENT
CONTENTCONTENTCONTENT</string>
</resources>
```

Strings extraction script from XML Translation file
```
keys=$(grep -oP "(?<=<string name=\").*?(?=\">)" strings.xml)
while read key_title
do

string_title=$(echo 'cat resources/string[@name="'${key_title}'"]/text()' | xmllint --shell strings.xml | sed -e '$d' -e '1d')

done < <(echo "$keys" | grep -P ".*(?===title)")

while read key_content
do

string_content=$(echo 'cat resources/string[@name="'${key_content}'"]/text()' | xmllint --shell strings.xml | sed -e '$d' -e '1d')

done < <(echo "$keys" | grep -Pv ".*(?===title)")
```

gnuhead-chieb commented:
@MuntashirAkon 
I've done most of process.Before I start finalization,Please check this PR to work script in your environment or not.

Please note that this script is require **bash**(Not work with sh) and **GNU Sed**(Not posix).
and you have to install or put [pandoc-crossref](https://github.com/lierdakil/pandoc-crossref/releases) binary into working directory(This binary isn't include in PR).Other dependencies is written in "doctool.sh help" option.

You can use strings_sampleja.xml for check translation process.

MuntashirAkon reviewed (commented):
A few questions for you:
1. Can we still build a pdf file? (It's not important, just for understanding)
2. Does every conversion process preserve the file structure i.e. is the output file identical upon subsequent build? This is very important otherwise we wouldn't be able reproduce the same binary file with the same set of configuration.
3. What's the state of the pdf files? Are the now converted into PNG or something?

PS: I haven't yet run them on my PC. I'll run them once you've updated the README to include how to install the dependencies in a standard path i.e. `/usr/bin/` or `/usr/local/bin`.

gnuhead-chieb commented:
>     1. Can we still build a pdf file?

No,To build pdf or something else, you have to convert to html once,then convert to else.



>     2\. Does every conversion process preserve the file structure?

Yes



>     3\. What's the state of the pdf files? Are the now converted into PNG or something?

These files are no meaning.I will remove them in finalizing process of this PR.

>     This should refer the list above, not the section.
I know about it.I'm looking for other implemention of this.

>     how the dependencies could be installed in both Linux and macOS. 
Most of these dependencies Just install by package manager.Then run required command on Terminal.It is OK if returned right output when run command on Terminal.

>     I don't know if xmllint is installed by default on Linux or not, macOS has it but it's way too outdated, you'd have to install it via Homebrew
xmllint is installed by default on most of Linux distros.This script doesn't using advanced feature of xmllint.so maybe works fine if it is outdated.

>     So, the only dependencies we should care about are pandoc-crossref,
If you want to install,
follow this guide
https://github.com/lierdakil/pandoc-crossref/blob/master/README.md

if not,
Download binary from [github releases](https://github.com/lierdakil/pandoc-crossref/releases) then put this on working directory(
AppManager/docs/raw/latex)

MuntashirAkon commented:
I try very hard to make the project fully reproducible to anybody. So, kindly add the requirements in the build instruction section in the README first so that no confusion arises. I know how to install pandoc or any other dependencies but I will follow the README myself to make sure that it's understandable to others.

gnuhead-chieb commented:
I added simple depschecker and install instruction（Debian based only,I dont use other distro or mac)

MuntashirAkon commented:
I don't see any modification of the README file.

Notice that the conversions will be done using an automated script like it is now when you run `./gradlew assembleDebug`. If a required binary cannot be found, it would simply throw error since the script would return non-zero code. You don't have to check for binaries there (but it's good that you've done as it can be helpful). But you need to add the instructions in the build instructions section of the README file. This way, one would only need read the build instructions to know what to do/install at a first glance.

gnuhead-chieb commented:
I added readme

MuntashirAkon commented:
We already have png images at `/docs/raw/images` folder, those files in the tex files if pandoc supports it instead of uploading the new png images.

MuntashirAkon commented:
Also, you are using GNU grep instead of the unix grep.

gnuhead-chieb commented:
> We already have png images at `/docs/raw/images` folder, those files in the tex files if pandoc supports it instead of uploading the new png images.

Yes.However I made this script work only latex folder.
You can clone only latex folder using this command
`svn checkout https://github.com/gnuhead-chieb/AppManager/branches/gnuhead-chieb-patch-1/docs/raw/latex`



> Also, you are using GNU grep instead of the unix grep.

Fixed

MuntashirAkon commented:
> Yes.However I made this script work only latex folder.

This should've very hard to configure.

Anyway, you've done a very good job, I think.

Here are my findings:

![Screenshot 2021-05-16 at 00 15 09](https://user-images.githubusercontent.com/8176128/118374114-2dc12f80-b5dc-11eb-997a-b906eaddad00.png)

We don't need bullet points, since we already have section numbering.

![Screenshot 2021-05-16 at 00 09 39](https://user-images.githubusercontent.com/8176128/118374135-4d585800-b5dc-11eb-84db-2075b9d60f75.png)

Figure numbers do not match. In the figure, it says Figure 1 but in the text it says Figure 7.1.

![Screenshot 2021-05-16 at 00 07 22](https://user-images.githubusercontent.com/8176128/118374151-65c87280-b5dc-11eb-97d3-e86da6d0dad1.png)

The links at the bottom should also be italicised. Use a common class name for the links if possible or make them bulleted.

![Screenshot 2021-05-16 at 00 07 04](https://user-images.githubusercontent.com/8176128/118374220-b770fd00-b5dc-11eb-8bfd-95fe070606df.png)

![Screenshot 2021-05-16 at 00 06 48](https://user-images.githubusercontent.com/8176128/118374229-bc35b100-b5dc-11eb-97c3-21942ba12b4b.png)

Use of non-standard colours for errors and warnings.

![Screenshot 2021-05-16 at 00 06 26](https://user-images.githubusercontent.com/8176128/118374246-cbb4fa00-b5dc-11eb-989c-b45b01f04831.png)

Remove the `>` sign. It looks odd and non-standard.

Apply the following changes to the `custom.css`:

```diff
diff --git a/docs/raw/latex/custom.css b/docs/raw/latex/custom.css
index a9446d74..396c8efb 100644
--- a/docs/raw/latex/custom.css
+++ b/docs/raw/latex/custom.css
@@ -35,7 +35,7 @@
     .main, .footnotes {max-width: 42em;}
 }
 
-.mdframed {
+.amalert {
     border-top: 0 !important;
     border-bottom: 0 !important;
     border-right: 0 !important;
@@ -44,6 +44,10 @@
     background: unset !important;
 }
 
+.amalert p:nth-child(1), .amalert p:nth-child(2) {
+    display: inline;
+}
+
 .itemize .colorbox {
     padding: 3px 7px;
     border-radius: 5px;
@@ -53,6 +57,10 @@ figcaption.caption span.id:after {
     content: ' ';
 }
 
+figure {
+    text-align: center;
+}
+
 @media (prefers-color-scheme: dark) {
     .sidenav {background: black !important;}
 
@@ -185,7 +193,6 @@ hr {
 }
 
 pre, code, kbd, samp {
-    color: #000;
     font-family: monospace, monospace;
     _font-family: 'courier new', monospace;
     font-size: 0.98em;

```

gnuhead-chieb commented:
> We don't need bullet points, since we already have section numbering.

Added todo list.I will check it later



> Figure numbers do not match. In the figure, it says Figure 1 but in the text it says Figure 7.1.

Added todo list



> The links at the bottom should also be italicised. Use a common class name for the links if possible or make them bulleted.

Fixed!



> Use of non-standard colours for errors and warnings.

These color codes are taken from build output from make4ht.If you have better color,Please specify it.



> Remove the `>` sign. It looks odd and non-standard.

Fixed!



Did you checked translation process?Does it worked correctly?
And for currently,translation merging process is very slow.If you know tricks to faster this code,please let me know.

gnuhead-chieb commented on docs/raw/latex/appendices/app_ops.tex:
```diff
@@ -132,7 +132,7 @@ \subsection{PackageOps}\label{subsec:package-ops}
 \end{listing}
 
-As can be seen in \Lref{lst:package-ops-class}, it stores all \textbf{OpEntry} for a package as well as the
+As can be seen in \Lref{subsec:package-ops}, it stores all \textbf{OpEntry} for a package as well as the
```
Fixed!

MuntashirAkon commented:
> These color codes are taken from build output from make4ht.If you have better color,Please specify it.

See here: https://muntashirakon.github.io/AppManager/en/#QQ2-1-42

> Did you checked translation process?Does it worked correctly?

They worked but for `faq_aot.tex` only.

> And for currently,translation merging process is very slow.If you know tricks to faster this code,please let me know.

I think most of the script can be rewritten using PHP. But don't worry about this right now.

gnuhead-chieb commented:
> See here: https://muntashirakon.github.io/AppManager/en/#QQ2-1-42

Color updated

> They worked but for `faq_aot.tex` only.

This is why,I have added stringkeys for this file only for now.I will add stringkeys for all tex files.
Multifile support in doctool.sh is already covered(And this is also reason of made translation merging process very slow)
So I only have to do for translation is add stringkeys for all tex files

MuntashirAkon commented:
> This is why,I have added stringkeys for this file only for now.I will add stringkeys for all tex files.
> Multifile support in doctool.sh is already covered(And this is also reason of made translation merging process very slow)
> So I only have to do for translation is add stringkeys for all tex files

I see. Please do add some comments on the script for clarity since it heavily uses Regex. I'll try to take care of the rests.

MuntashirAkon commented:
Please resolve the conflicts, and briefly explain what to test here for me before finalising the docs.

gnuhead-chieb commented:
I resolved conflicts.
Run commands wrote in [Readme](https://github.com/gnuhead-chieb/AppManager/blob/gnuhead-chieb-patch-1/docs/raw/latex/README)'s tutorial section and check if any bugs are appear.

gnuhead-chieb commented:
added

MuntashirAkon commented:
@gnuhead-chieb: Can you write a script (simple bash script, for example) which converts a sample TEX file to XML and converts the XML back to TEX and show that both input and output files matched? The sample TEX file should consist of a few Latex commands as well as characters or symbols that needs conversion such as `<` (greater than), `>` (less than) and `&` (ampersand). Along with the sample TEX file, we will also test all the tex files located in the latex folder. I will include the script in tests so that we can be assured that new additions to the tex files aren't breaking anything.

gnuhead-chieb commented:
> Can you write a script (simple bash script, for example) which converts a sample TEX file to XML and converts the XML back to TEX and show that both input and output files matched?

Remove `while read VARIABLE 〜 done < <(COMMAND)` on conversion process and replace variable to sample file's name.
Then just compare both files by command like `[[ $(cat INPUT) = $(cat OUTPUT) ]] && echo OK || echo Files are different`

MuntashirAkon commented:
Well, after running `./doctool.sh updatetranslation` and `./doctool.sh mergetranslation`, the `.tex` files in the _latextranslated_ folder is expected to have the exact same contents as in the _latex_ folder. But it doesn't seem so. Some sections appear to be missing.

gnuhead-chieb commented:
> Well, after running `./doctool.sh updatetranslation` and `./doctool.sh mergetranslation`, the `.tex` files in the _latextranslated_ folder is expected to have the exact same contents as in the _latex_ folder. But it doesn't seem so. Some sections appear to be missing.

fixed

MuntashirAkon commented:
Some characters might not be translated correctly. For example, `&` is translated as `&amp;`.

gnuhead-chieb commented:
fixed

MuntashirAkon commented:
`appendices/app_ops.tex` still didn't match. At this point, use `sha*sum` (linux) or `shasum -a *` (macOS) to make sure that output tex matches with inputs. This should ensure that both the outputs are correctly being produced, and nothing is skipped. Did you check if Japanese texts are working correctly? Also, remove verbose logging from the `updatetranslation` (if not necessary) as printing makes operation a bit slow.

gnuhead-chieb commented:
> appendices/app_ops.tex still didn't match. 

It seems the problem is what sed couldn't handle nested curly braces（{ }）
As written in readme(and I said before),this script extract title strings from `section{,subsection{,subsubsection{,chapter{,caption{,paragraph{` to next `}`.
appendices/app_ops.tex uses nested braces.
so it will cause bug.
ex)`\subsection{\texttt{OP\_*} Constants}`

But it has problem that it is difficult to handle nested string with regex.
If you know better way to handle nested braces and extract it, please let me know.

gnuhead-chieb commented:
> Did you check if Japanese texts are working correctly?

Yes,but only for `faq/aot.tex`.I only translated this part right now.

> Also, remove verbose logging from the updatetranslation (if not necessary) as printing makes operation a bit slow.
Ok,I7ll remove it

gnuhead-chieb commented:
For now,every file will be identical probably.
but nesting problem wasn't fixed yet.
Also,"cd $0" was disabled because it seems not working at all.

MuntashirAkon commented:
> but nesting problem wasn't fixed yet.

What nesting problem?

Apart from other issues specified earlier, I've found a few minor issues:

1. There should be margin between two consecutive tips/note/warnings.
   ![Screenshot 2021-11-20 at 10 56 32](https://user-images.githubusercontent.com/8176128/142760660-78d91a5e-e92b-4741-93af-48045c2fc052.png)
2. The see also section (not the inline ones) should contain bullet points
   ![Screenshot 2021-11-20 at 10 36 00](https://user-images.githubusercontent.com/8176128/142760701-cf1b9b52-784d-4d95-af2d-2aade44d6098.png)
3. The stylesheet needs some changes in order to make the UI identical to the make4ht. We only need to edit custom.css, I believe.

gnuhead-chieb commented:
> What nesting problem?

I wrote here
https://github.com/MuntashirAkon/AppManager/pull/418#issuecomment-974797406
and see string `appendices_appops-op-constants==title`

> The stylesheet needs some changes in order to make the UI identical to the make4ht. We only need to edit custom.css, I believe.
Yeah, but I'm not very familiar with CSS.I will investigate and fix them if I can. if not,please change them after this PR merged

MuntashirAkon commented:
> I wrote here
> [#418 (comment)](https://github.com/MuntashirAkon/AppManager/pull/418#issuecomment-974797406)
> and see string `appendices_appops-op-constants==title`

I see. I think this is because regex cannot understand semantics, it can only be used for parsing. You'd need to some sort of memory elements (i.e. stacks) in order to keep track of the brackets, but that would make things a lot complicated (may be there's a antlr alternative in Python). I wonder why there's not a single project with proper support for localisation. At this point, I still maintain that localisation after converting latex to HTML is the best option available for us. I would've written a project myself if I had enough time after all the things that I do.

gnuhead-chieb commented:
I will extract all title strings instead of extract strings in section tag where has nesting problem.(`%%#*string_key==titlex>>`)

ex) 
`\subsection{\texttt{OP\_*} Constants}\label{subsec:op-constants} %%#*appendices_appops-op-constants==titlex>>` 
will extracted as:
`<string name="appendices_changelogs-chapter==title"><![CDATA[\subsection{\texttt{OP\_*} Constants}\label{subsec:op-constants}]]></string>` 
in xliff.

> maintain that localisation after converting latex to HTML is the best option available for us

No,it more worse than this problem.Due to no string key,it will break all strings every resource update.Also,It hasn't more controls like option to choose not-translating strings.(see string intro_main-donation)

gnuhead-chieb commented:
oh i found solution.nesting problem fixed

MuntashirAkon commented:
> Due to no string key,it will break all strings every resource update.

Ah, I see.

gnuhead-chieb commented:
> There should be margin between two consecutive tips/note/warnings.
fixed

gnuhead-chieb commented:
done

gnuhead-chieb commented:
@MuntashirAkon Please check last commit

gnuhead-chieb commented:
is git rebase -i HEAD~n can squash commits?
